chore(cd): fix cd + setup npm provenance

louis-bompart opened the pull request:
- Fix CD by ensuring we checkout first.
- Setup NPM Provenance See [NPM Doc](https://docs.npmjs.com/generating-provenance-statements#example-github-actions-workflow)
- Update all reusable actions to their latest versions
- Resolve permission issue for `dependency-review`

louis-bompart committed:
chore: add npm provenance

louis-bompart committed:
+fix: checkout before running cd

louis-bompart commented on .github/workflows/cd.yml:
```diff
@@ -17,7 +21,7 @@ jobs:
           NODE_AUTH_TOKEN: ${{ secrets.NPM_TOKEN }}
-  
\ No newline at end of file
+  
```
```suggestion
```

louis-bompart committed:
clean eol

samisayegh commented on .github/workflows/cd.yml:
```diff
@@ -8,7 +8,11 @@ jobs:
     name: CD
     runs-on: ubuntu-latest
     environment: CD
+    permissions:
+      contents: read
+      id-token: write
```
Why is this block needed?

louis-bompart replied:
The content read, well it doesn't do much, just allow a read access to the commits etc.
The id-token is a requirement for the provenance signature.

samisayegh commented on .github/workflows/cd.yml:
```diff
@@ -8,7 +8,11 @@ jobs:
+      contents: read
+      id-token: write
     steps:
+      - uses: actions/checkout@v4
```
Interesting that this step was missing hmm. Was the `CD` workflow just not working?
If we're bumping to v4, could we do the same for the `CI` workflow too?

louis-bompart replied:
- [x] update actions checkout in CI.yml

louis-bompart committed:
fix: update GitHub Actions to use checkout and setup-node v4

louis-bompart committed:
fix(dep-review): add missing permission

The workflow is not valid. .github/workflows/dependency-review.yml (Line: 10, Col: 3): Error calling workflow 'coveo/public-actions/.github/workflows/dependency-review.yml@main'. The nested job 'dependency-review' is requesting 'pull-requests: write', but is only allowed 'pull-requests: none'.
